File: 二叉树/第二周/257. 二叉树的所有路径.js
var binaryTreePaths = function (root) {
    let totalPaths = [];
    let currentPath = [root.val];

    traversal(root, currentPath, totalPaths);
    return totalPaths;
};

/*
 * 1. 递归函数函数参数以及返回值
 * 参数只需要根节点，不需要返回值，函数调用过程中负责将目前的节点加入 currentPath 并使其左右子节点参与构建路径
 *
 * 2. 确定递归终止条件
 * 到叶子节点时，就开始结束的处理逻辑（把路径放进result里）。
 * 那么什么时候算是找到了叶子节点？ 是当 cur不为空，其左右孩子都为空的时候，就找到叶子节点。
 * 所以本题的终止条件是：if (node.left === null && node.right === null)
 *
 * 3. 确定单层递归逻辑
 * 因为是前序遍历，需要先处理中间节点，这里对中间节点的处理也是判断它是否是叶子节点(终止条件).
 *
 * 然后就是递归和回溯的过程。对节点的左右子树进行递归从而构建路径的过程中，因为 currentPath 不能一直加入节点，它还要删节点，然后才能加入新的节点。
 *
 */

function traversal(node, currentPath, totalPaths) {
    if (node.left === null && node.right === null) {
        let pathStr = currentPath.join("->");
        totalPaths.push(pathStr);
        return;
    }

    //* 对左子树进行递归
    if (node.left) {
        currentPath.push(node.left.val);
        //# 回溯和递归是一一对应的，有一个递归，就要有一个回溯
        traversal(node.left, currentPath, totalPaths);
        currentPath.pop();
    }

    //* 对右子树进行递归
    if (node.right) {
        currentPath.push(node.right.val);
        //# 回溯和递归是一一对应的，有一个递归，就要有一个回溯
        traversal(node.right, currentPath, totalPaths);
        currentPath.pop();
    }
}

File: 二叉树/第一周(二叉树的遍历)/257. 二叉树的所有路径.py

""" 
* 涉及到回溯，因为我们要把路径(currentPath)记录下来,需要回溯来回退一个路径,再进入另一个路径。
* 回溯和递归是一一对应的，有一个递归，就要有一个回溯. 并且它们要放在同一个括号里
/ People who are saying time complexity is O(n) are wrong IMHO. Yes it is true that you need to visit every node at least once, so that contributes O(n) to the total time complexity. But then at the same time when you reach the leaf, you need to generate the path and add it to the ans array, which will be O(logN) time in the case when tree has some leaves on both sides. And when the tree is somewhat balanced, you can take the number of leaves as half of the number of nodes in the tree, as an upper bound. So, that gives you O(n/2 + (n/2)logn) time which turns out to be O(nlogn).
/ Now consider the case where the tree is unbalanced. So, all nodes are on one side like a list. That will give you a height of O(n) but at the same time only one leaf will be present there. So it will give the time complexity as O(n-1 + n), which is O(n). So, the case which turns out to be the worst generally for most tree problems is not the worst case here.
/ By this we can conclude that as the number of leaves increase in the tree, the time complexity moves towards the worst side, as more paths will be generated.
? https://leetcode.com/problems/binary-tree-paths/discuss/68258/Accepted-Java-simple-solution-in-8-lines/562740
"""


class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        if root is None:
            return []
        result = []
        curPath = [str(root.val)]

        def backtracking(root):
            if root.left is None and root.right is None:
                result.append('->'.join(curPath))
                return

            if root.left is not None:
                curPath.append(str(root.left.val))
                backtracking(root.left)
                curPath.pop()

            if root.right is not None:
                curPath.append(str(root.right.val))
                backtracking(root.right)
                curPath.pop()

        backtracking(root)

        return result
